File: take_screenshots.py
import os
import time
import re
import argparse
from playwright.sync_api import sync_playwright

def main():
    parser = argparse.ArgumentParser(description="Capture advanced charts screenshots.")
    parser.add_argument("--portfolio", "-p", default="portfolio-01.csv", help="Portfolio CSV filename to analyze")
    args = parser.parse_args()
    
    portfolio_file = args.portfolio
    print(f"Target Portfolio file: {portfolio_file}")

    output_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "frontend", "public", "images_advanced")
    os.makedirs(output_dir, exist_ok=True)
    
    with sync_playwright() as p:
        print("Launching browser...")
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.set_viewport_size({"width": 1600, "height": 1200})
        
        print("Navigating to http://localhost:3000...")
        try:
            page.goto("http://localhost:3000", wait_until="networkidle", timeout=30000)
        except Exception as e:
            print("Navigation load timeout/error, continuing: ", e)
            
        time.sleep(3)
        
        # 1. Change portfolio filename to dynamic target
        print(f"Selecting portfolio '{portfolio_file}'...")
        page.wait_for_selector("select")
        page.select_option("select", portfolio_file)
        
        # Wait for loading batch analysis to complete
        print("Waiting for portfolio batch analysis to complete...")
        try:
            page.wait_for_selector("text=Adv. Charts", timeout=90000)
            print("Dashboard layout loaded.")
        except Exception as e:
            print("Timeout waiting for Adv. Charts button to appear: ", e)
        
        time.sleep(3) # Let UI stabilize
        
        # 2. Select the "Adv. Charts" tab button
        print("Switching to Adv. Charts tab...")
        adv_charts_btn = None
        try:
            # First try matching exact text
            adv_charts_btn = page.get_by_text("Adv. Charts").first
            adv_charts_btn.click()
            print("Switched to Adv. Charts view via get_by_text.")
        except Exception as e:
            print("Failed to click Adv. Charts using locator: ", e)
            buttons = page.query_selector_all("button")
            for btn in buttons:
                try:
                    text = btn.inner_text() or ""
                    if "Adv. Charts" in text or "Adv" in text:
                        adv_charts_btn = btn
                        btn.click()
                        print(f"Clicked button with text '{text}' via fallback search.")
                        break
                except Exception:
                    pass
                
        if not adv_charts_btn:
            print("Could not find Adv. Charts button!")
            browser.close()
            return
            
        time.sleep(4)
        
        # 3. Toggle off unwanted indicators
        print("Toggling chart overlays to keep only required ones...")
        items_to_toggle_off = [
            'SMA 9', 'SMA 12', 'SMA 26', 'SMA 50', 'SMA 200',
            'BB Upper', 'BB Lower', 'SMA 20 (BB)'
        ]
        
        plot_buttons = page.query_selector_all("button")
        for btn in plot_buttons:
            try:
                text = btn.inner_text().strip()
                if text in items_to_toggle_off:
                    print(f"Toggling off {text}...")
                    btn.click()
                    time.sleep(0.3)
            except Exception as e:
                pass
                
        print("Indicators isolated successfully.")
        
        # 4. Retrieve all stock symbols currently in the sidebar
        tickers = []
        divs = page.query_selector_all("div")
        for div in divs:
            try:
                class_name = div.get_attribute("class") or ""
                if "cursor-pointer" in class_name:
                    span = div.query_selector("span")
                    if span:
                        symbol_text = span.inner_text().strip()
                        if symbol_text and re.match(r"^[A-Z0-9.\-]+$", symbol_text) and len(symbol_text) <= 10:
                            if symbol_text not in tickers:
                                tickers.append(symbol_text)
            except Exception:
                pass
                
        print(f"Found tickers in sidebar: {', '.join(tickers)}")
        
        # 5. Loop through each ticker, click, wait, and capture screenshot of advanced chart card
        for symbol in tickers:
            print(f"Processing {symbol}...")
            
            # Find the sidebar element for this symbol and click it
            target_el = None
            sidebar_divs = page.query_selector_all("div")
            for div in sidebar_divs:
                try:
                    class_name = div.get_attribute("class") or ""
                    if "cursor-pointer" in class_name:
                        span = div.query_selector("span")
                        if span and span.inner_text().strip() == symbol:
                            target_el = div
                            break
                except Exception:
                    pass
                    
            if target_el:
                print(f"Clicking sidebar for {symbol}...")
                target_el.click()
                time.sleep(5) # Wait for API and recharts animation
                
                # Find the card element for Advanced Charts and take screenshot
                all_divs = page.query_selector_all("div")
                chart_card = None
                for div in all_divs:
                    try:
                        inner_text = div.inner_text() or ""
                        if "Price Action & Moving Averages" in inner_text and ("Daily Close with SMAs" in inner_text or "Close Price" in inner_text):
                            chart_card = div
                            break
                    except Exception:
                        pass
                        
                if chart_card:
                    screenshot_path = os.path.join(output_dir, f"{symbol}_advanced_chart.png")
                    print(f"Saving screenshot of chart card to {screenshot_path}...")
                    chart_card.screenshot(path=screenshot_path)
                else:
                    print(f"Could not find chart card for {symbol}, taking full page screenshot...")
                    screenshot_path = os.path.join(output_dir, f"{symbol}_full_page.png")
                    page.screenshot(path=screenshot_path)
            else:
                print(f"Could not find sidebar item for {symbol}")
                
        print("Screenshot generation complete.")
        browser.close()

if __name__ == "__main__":
    main()

File: take_screenshots.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

(async () => {
  const outputDir = path.join(__dirname, 'images_advanced');
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  console.log('Launching browser with native 80% window zoom...');
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--force-device-scale-factor=0.8']
  });

  const page = await browser.newPage();
  await page.setViewport({ width: 1600, height: 1200 });

  console.log('Navigating to http://localhost:3000...');
  try {
    await page.goto('http://localhost:3000', { waitUntil: 'networkidle2', timeout: 30000 });
  } catch (err) {
    console.log('Networkidle2 timed out, proceeding anyway...');
  }

  // 1. Change portfolio filename to 'portfolio-01.csv'
  console.log("Selecting portfolio 'portfolio-01.csv'...");
  await page.waitForSelector('select');
  await page.select('select', 'portfolio-01.csv');

  // Wait for loading indicator or load completion
  console.log('Waiting for portfolio batch analysis to complete...');
  await page.waitForFunction(() => {
    return !document.body.innerText.includes('Running Batch Quant Models');
  }, { timeout: 90000 });

  await delay(4000); // Let UI stabilize

  // 2. Select the "Adv. Charts" tab button
  console.log('Switching to Adv. Charts tab...');
  const buttons = await page.$$('button');
  let advChartsBtn = null;
  for (const btn of buttons) {
    const text = await page.evaluate(el => el.textContent, btn);
    if (text.includes('Adv. Charts')) {
      advChartsBtn = btn;
      break;
    }
  }

  if (advChartsBtn) {
    await advChartsBtn.click();
    console.log('Switched to Adv. Charts view.');
  } else {
    console.error('Could not find Adv. Charts button!');
    await browser.close();
    return;
  }

  await delay(3000);

  // 3. Toggle off all indicators except Close Price, Willy VWAP, VWAP Upper, VWAP Lower
  console.log('Toggling chart overlays to keep only required ones...');
  const itemsToToggleOff = [
    'SMA 9',
    'SMA 12',
    'SMA 26',
    'SMA 50',
    'SMA 200',
    'BB Upper',
    'BB Lower',
    'SMA 20 (BB)'
  ];

  // We loop a few times to make sure we find all and click them
  let plotButtons = await page.$$('button');
  for (const btn of plotButtons) {
    try {
      const text = await page.evaluate(el => el.textContent.trim(), btn);
      if (itemsToToggleOff.includes(text)) {
        console.log(`Toggling off ${text}...`);
        await btn.click();
        await delay(200);
      }
    } catch (e) {
      // Element might detach, ignore
    }
  }

  console.log('Indicators isolated successfully.');

  // 4. Retrieve all stock symbols currently in the sidebar
  const tickers = await page.evaluate(() => {
    const items = Array.from(document.querySelectorAll('div'));
    const symbols = [];
    for (const item of items) {
      if (item.className.includes('cursor-pointer') && item.querySelector('span')) {
        const symbolText = item.querySelector('span').textContent.trim();
        if (symbolText && /^[A-Z0-9.\-]+$/.test(symbolText) && symbolText.length <= 10) {
          if (!symbols.includes(symbolText)) {
            symbols.push(symbolText);
          }
        }
      }
    }
    return symbols;
  });

  console.log(`Found tickers in sidebar: ${tickers.join(', ')}`);

  // 5. For each ticker, select it, wait, and capture a screenshot of the Advanced Charts card
  for (const symbol of tickers) {
    console.log(`Processing ${symbol}...`);
    
    // Find the sidebar element for this symbol and click it
    const sidebarElements = await page.$$('div');
    let targetEl = null;
    for (const el of sidebarElements) {
      try {
        const text = await page.evaluate(element => {
          const span = element.querySelector('span');
          return span ? span.textContent.trim() : null;
        }, el);
        if (text === symbol && await page.evaluate(element => element.className.includes('cursor-pointer'), el)) {
          targetEl = el;
          break;
        }
      } catch (e) {}
    }

    if (targetEl) {
      console.log(`Clicking sidebar for ${symbol}...`);
      await targetEl.click();
      await delay(4000); // Wait for API and chart lines to render

      // Find the container element for all Advanced Charts and take an element screenshot
      const cards = await page.$$('div');
      let chartCard = null;
      for (const card of cards) {
        try {
          const isTarget = await page.evaluate(element => {
            const className = element.className || '';
            const text = element.innerText || '';
            return className.includes('space-y-6') && 
                   !className.includes('max-w-7xl') && 
                   text.includes('Price Action & Moving Averages') && 
                   text.includes('MACD');
          }, card);
          if (isTarget) {
            chartCard = card;
            break;
          }
        } catch (e) {}
      }

      if (chartCard) {
        const screenshotPath = path.join(outputDir, `${symbol}_advanced_chart.png`);
        console.log(`Saving screenshot to ${screenshotPath}...`);
        await chartCard.screenshot({ path: screenshotPath });
      } else {
        console.warn(`Could not find chart card for ${symbol}, taking viewport screenshot...`);
        const screenshotPath = path.join(outputDir, `${symbol}_full_viewport.png`);
        await page.screenshot({ path: screenshotPath });
      }
    } else {
      console.error(`Could not find sidebar item for ${symbol}`);
    }
  }

  console.log('Screenshot generation complete.');
  await browser.close();
})();
